doc/dev/kdb-operations.md

```diff
@@ -0,0 +1,156 @@
+- explicit `default:/` mountpoints are illegal
+  They can only exist via cascading mountpoints.
+  This is so that default values can be checked and validated by plugins.
+- `spec:/` mountpoints may exist, but they are treated differently during `get` and `set`
+- `proc:/` mountpoints may exist, but they are read-only (see below)
+- `dir:/`, `user:/` and `system:/` mountpoints can be created without restrictions
+
+Other restrictions:
+
+- Creating mountpoints for root keys (`system:/`, `user:/`, etc.) is not allowed.
+  These parts of the KDB always use hardcoded backends that can only be configured at compile-time.
+- Creating a mountpoint for `system:/elektra` or any mountpoint below is not allowed.
+  This part of the KDB will always be stored in the bootstrap backend.
+  While the only part that is actually required for the bootstrap process is the one below `system:/elektra/mountpoints`, everything below `system:/elektra` is reserved and should only be used internally.
+  Additionally, `system:/elektra/version` and `system:/elektra/modules` will always point to hardcoded read-only backends containing information about this Elektra installation.
+
+## `get` Operation
+
+The purpose of the `get` operation is to read data stored in backends into a `KDB` instance.
+
+> **Note:** Some details of a `get` operation are defined in the [contract with backend plugins](./backend-plugins.md).
+
+Properties of `kdbGet()`:
+
+- After calling `kdbGet (kdb, ks, parentKey)`, the KeySet `ks` will contain _all keys_ (including their values) that are stored in _any backend_ with a mountpoint that is _below `parentKey`_.
+- After calling `kdbGet (kdb, ks, parentKey)`, below `parentKey` the KeySet `ks` will _only_ contain keys that are stored in a backend.
+- The KeySet `ks` may contain other keys not below `parentKey`.
+  These keys fall into one of two categories:
+  1. Keys that are stored in a backend that is not below `parentKey`.
+     `kdbGet()` may decide that it is more efficient (e.g. because of a cache) to return more keys than requested.
+  2. Keys that were already present in `ks` when `kdbGet()` was called.
+
+In simpler terms, this means to the caller it looks as if `kdbGet()` had removed all keys below `parentKey` from `ks` and then loaded the data from the backends.
+Below `parentKey` the KeySet `ks` correctly represents the state of the KDB.
+For the rest of `ks` there are no such guarantees.
+
+TODO: cache correct? notifications? (just call at the end?)
+
+The basic flow of this operation is:
+
+1. Determine the backends needed to read all keys below `parentKey`
+2. Run the `init` phase on all the backends that haven't been initialized
+3. Run the `resolver` phase on all backends
+4. From now on ignore all backends, which indicated that there is no update.
+5. If all backends are now ignored, **return**.
+6. If a global cache plugin is enabled:
+   Ask the global cache plugin for the cache times for all backends.
```

Comment:
No, this part compares whether all parentKeys match  the cached versions etc.

Currently the resolver compares the cached times / mtimes and returns this to the core:
https://github.com/ElektraInitiative/libelektra/blob/a80240f20c735514d9ba4f7a98a1306f11eab12d/src/libs/elektra/kdb.c#L673-L675

I'm also quite confident that the resolvers are the best place to do this comparison. It's not clear that mtime is the only possibility: e.g. for gitresolver the actual change may be detected using commit hash or so.

Reply:
> Currently the resolver compares the cached times / mtimes and returns this to the core

Yes, the comparison doesn't happen here. (see also https://github.com/ElektraInitiative/libelektra/pull/3693#discussion_r607113417) But the `resolver` plugin just reads from the global keyset. Where is the code that loads the cached times into the global keyset? That's what I thought was happening here (among other things).

> I'm also quite confident that the resolvers are the best place to do this comparison. It's not clear that mtime is the only possibility: e.g. for gitresolver the actual change may be detected using commit hash or so.

See https://github.com/ElektraInitiative/libelektra/pull/3693#discussion_r607933606.

Reply:
Ah, the cached times are actually loaded by mmapstorage into the global keyset.

Since the global keyset is just a keyset it is simply also cached and then loaded here:

https://github.com/ElektraInitiative/libelektra/blob/5702ce0496823fdd12c4e1e6ff9d6d1ca1b71a40/src/plugins/mmapstorage/mmapstorage.c#L1147-L1161

Reply:
Ah, so the `cache` plugin writes the times into the global keyset when the cache is updated. This is stored by `mmapstorage` and when the cache is loaded everything is there already. That would actually be very helpful, then this

https://github.com/ElektraInitiative/libelektra/blob/898990819649541dafeca1b74717e4e1ba24e1e1/doc/dev/backend-plugins.md#L84-L85

wouldn't be needed and format wouldn't be a problem.

Reply:
> Ah, so the cache plugin writes the times into the global keyset when the cache is updated.

No, the resolver puts the times into the global keyset when a file is read. Resolver is the only code which knows what a change is (in this case mtime), and decides about all of this.

Putting into global keyset happens here:
https://github.com/ElektraInitiative/libelektra/blob/6b228bfca678dcc4a0dcf08adb2ae5e28cef278c/src/plugins/resolver/resolver.c#L603-L609

Reply:
> No, the resolver puts the times into the global keyset when a file is read. Resolver is the only code which knows what a change is (in this case mtime), and decides about all of this.

Okay that works too. Then we could actually go back to my original concept of using "cache entry IDs" instead of "cache terms". That might work better for e.g. a git-based backend were you could use a commit hash instead of a timestamp.